### PR TITLE
Course Aggregate updater flink job

### PR DESCRIPTION
**Pull request by manjudr**
Design wiki: https://project-sunbird.atlassian.net/wiki/spaces/SBDES/pages/1520042020/Course+Infra+-+Async+Jobs+-+Implementation+Design

### CODE REVIEW

File: core-job/pom.xml [hunk 1]
```diff
@@ -4,15 +4,15 @@
...
     <modelVersion>4.0.0</modelVersion>
     <parent>
-        <groupId>org.sunbird.kp</groupId>
+        <groupId>org.sunbird</groupId>
         <artifactId>knowledge-platform-jobs</artifactId>
         <version>1.0</version>
     </parent>
 
-    <artifactId>kp-jobs-core</artifactId>
+    <artifactId>core-job</artifactId>
```
Comment (maheshkumargangula) [inline core-job/pom.xml:12]:
```suggestion
    <artifactId>jobs-core</artifactId>
```

File: core-job/src/main/scala/org/sunbird/async/core/cache/DataCache.scala [hunk 2]
```diff
@@ -0,0 +1,156 @@
+package org.sunbird.async.core.cache
```
Comment (maheshkumargangula) [inline core-job/src/main/scala/org/sunbird/async/core/cache/DataCache.scala:1]: Remove unused methods.

File: course-aggregate-updater/src/main/resources/course-aggregate-updater.conf [hunk 3]
```diff
@@ -0,0 +1,29 @@
+include "base-config.conf"
+
+kafka {
+  input.topic = "sunbirddev.coursebatch.job.request"
+  output.audit.topic = "flink.kp.audit_event"
```
Comment (maheshkumargangula) [inline course-aggregate-updater/src/main/resources/course-aggregate-updater.conf:5]:
```suggestion
  output.audit.topic = "sunbirddev.telemetry.raw"
```

File: course-aggregate-updater/src/main/resources/course-aggregate-updater.conf [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+
+lms-cassandra {
+  keyspace = "sunbird_courses"
+  consumption.table = "user_content_consumption"
+  activity_user_agg.table = "activity_user_agg"
+}
+
+redis {
+  database {
+    nodes.id = 10
```
Comment (maheshkumargangula) [inline course-aggregate-updater/src/main/resources/course-aggregate-updater.conf:24]:
```suggestion
    relationCache.id = 10
```

File: course-aggregate-updater/src/main/scala/org/sunbird/async/functions/ProgressUpdater.scala [hunk 5]
```diff
@@ -0,0 +1,319 @@
...
+import org.sunbird.async.core.cache.{DataCache, RedisConnect}
+import org.sunbird.async.core.job.{Metrics, WindowBaseProcessFunction}
+import org.sunbird.async.core.util.CassandraUtil
+import org.sunbird.async.domain._
+import org.sunbird.async.task.CourseAggregateUpdaterConfig
+
+import scala.collection.JavaConverters._
+import scala.collection.mutable.ListBuffer
+
+class ProgressUpdater(config: CourseAggregateUpdaterConfig)(implicit val stringTypeInfo: TypeInformation[String], @transient var cassandraUtil: CassandraUtil = null) extends WindowBaseProcessFunction[util.Map[String, AnyRef], String, String](config) {
```
Comment (maheshkumargangula) [inline course-aggregate-updater/src/main/scala/org/sunbird/async/functions/ProgressUpdater.scala:25]:
```suggestion
class CourseAggregatesFunction(config: CourseAggregateUpdaterConfig)(implicit val stringTypeInfo: TypeInformation[String], @transient var cassandraUtil: CassandraUtil = null) extends WindowBaseProcessFunction[util.Map[String, AnyRef], String, String](config) {
```

File: course-aggregate-updater/src/main/scala/org/sunbird/async/functions/ProgressUpdater.scala [hunk 6]
```diff
@@ -0,0 +1,319 @@
...
+    cache = new DataCache(config, new RedisConnect(config), config.nodeStore, List())
+    cache.init()
+  }
+
+  override def close(): Unit = {
+    super.close()
+  }
+
+  def process(key: String, context: ProcessWindowFunction[util.Map[String, AnyRef], String, String, TimeWindow]#Context, events: lang.Iterable[util.Map[String, AnyRef]], metrics: Metrics): Unit = {
+    val batchEnrolmentEvents = events.asScala.filter(event => event.get(config.eData).asInstanceOf[util.Map[String, AnyRef]].asScala(config.action).asInstanceOf[String] == config.batchEnrolmentUpdateCode)
```
Comment (maheshkumargangula) [inline course-aggregate-updater/src/main/scala/org/sunbird/async/functions/ProgressUpdater.scala:49]:
```suggestion
    val batchEnrolmentEvents = events.asScala.filter(event => StringUtils.equals(event.get(config.eData).asInstanceOf[util.Map[String, AnyRef]].asScala(config.action).asInstanceOf[String], config.batchEnrolmentUpdateCode))
```

File: course-aggregate-updater/src/main/scala/org/sunbird/async/functions/CourseAggregatesFunction.scala [hunk 7]
```diff
@@ -0,0 +1,364 @@
...
+        ContentStatus(x._1, x._2.intValue(), completedCount)
+      }).groupBy(f => f.contentId)
+
+    enrichedContents.map(content => {
+      val consumedList = content._2
+      val finalStatus = consumedList.map(x => x.status).max
+      val views = sumFunc(consumedList, (x: ContentStatus) => {x.viewCount})
+      val completion = sumFunc(consumedList, (x: ContentStatus) => {x.completedCount})
+      (content._1, ContentStatus(content._1, finalStatus, completion, views))
+    }).toMap
```
Comment (SanthoshVasabhaktula) [inline course-aggregate-updater/src/main/scala/org/sunbird/async/functions/CourseAggregatesFunction.scala:296]: As discussed we don't need to compute viewCount. completedCount can be considered as viewCount

File: course-aggregate-updater/src/main/scala/org/sunbird/async/functions/CourseAggregatesFunction.scala [hunk 8]
```diff
@@ -0,0 +1,364 @@
...
+
+    // content status from the telemetry event.
+    val inputUserConsumptionList: List[UserContentConsumption] = groupedData.map(ed => {
+      val userConsumedContents = ed(config.contents).asInstanceOf[List[Map[String, AnyRef]]]
+      val enrichedContents = getContentStatusFromEvent(userConsumedContents)
+      UserContentConsumption(ed(config.userId).asInstanceOf[String], ed(config.batchId).asInstanceOf[String], ed(config.courseId).asInstanceOf[String], enrichedContents)
+    })
+
+    // Fetch the content status from the table in batch format
+    val dbUserConsumption: Map[String, UserContentConsumption] = getContentStatusFromDB(eDataBatch, metrics).map(e => (getUCKey(e), e)).toMap
```
Comment (SanthoshVasabhaktula) [inline course-aggregate-updater/src/main/scala/org/sunbird/async/functions/CourseAggregatesFunction.scala:66]: This data is going to be huge - right? I think we should make a read query per user?

File: course-aggregate-updater/src/main/scala/org/sunbird/async/functions/CourseAggregatesFunction.scala [hunk 9]
```diff
@@ -0,0 +1,364 @@
...
+   * Course Level Agg using the merged data of ContentConsumption per user, course and batch.
+   * @param userConsumption
+   * @param metrics
+   * @return
+   */
+  def courseActivityAgg(userConsumption: UserContentConsumption)(implicit metrics: Metrics): UserActivityAgg = {
+    val courseId = userConsumption.courseId
+    val userId = userConsumption.userId
+    val contextId = "cb:"+ userConsumption.batchId
+    val leafNodes = readFromCache(key = s"$courseId:$courseId:${config.leafNodes}", metrics).distinct
```
Comment (SanthoshVasabhaktula) [inline course-aggregate-updater/src/main/scala/org/sunbird/async/functions/CourseAggregatesFunction.scala:107]: Job should stop if the cache hit is a miss

Comment (maheshkumargangula) [reply course-aggregate-updater/src/main/scala/org/sunbird/async/functions/CourseAggregatesFunction.scala:110]: @manjudr - this is done.

File: course-aggregate-updater/src/main/scala/org/sunbird/async/functions/CourseAggregatesFunction.scala [hunk 10]
```diff
@@ -0,0 +1,364 @@
...
+
+  override def open(parameters: Configuration): Unit = {
+    super.open(parameters)
+    cassandraUtil = new CassandraUtil(config.dbHost, config.dbPort)
+    cache = new DataCache(config, new RedisConnect(config), config.nodeStore, List())
+    cache.init()
+  }
+
+  override def close(): Unit = {
+    super.close()
```
[the remark this answers is not in the document]
Comment (maheshkumargangula) [reply course-aggregate-updater/src/main/scala/org/sunbird/async/functions/CourseAggregatesFunction.scala:46]: @manjudr - this is done.

**Pull request merged**